libr/core/canal.c

```diff
@@ -2753,39 +2753,76 @@ R_API void r_core_recover_vars(RCore *core, RAnalFunction *fcn, bool argonly) {
 	return;
 }
 
-R_API RList* r_core_anal_graph_to(RCore *core, ut64 addr, int n) {
-	RAnalBlock *bb, *root = NULL, *dest = NULL;
-	RListIter *iter, *iter2;
-	RList *list2 = NULL, *list = NULL;
-	RAnalFunction *fcn;
+static bool anal_path_exists(RCore *core, ut64 from, ut64 to, RList *bbs, int depth, RList *state) {
+	RAnalBlock *bb = r_anal_bb_from_offset (core->anal, from);
+	RListIter *iter = NULL;
+	RList *refs = NULL;
+	RAnalRef *refi;
+	RAnalFunction *fcn = NULL;
+	RAnalFunction *cur_fcn = NULL;
+
+	if (depth < 1) {
+		eprintf ("going too deep\n");
+		return false;
+	}
 
+	if (!bb || !bbs) {
+		return false;
+	}
+	r_list_append(state, (void *)from);
+
+	// try to find the target in the current function
+	if (r_anal_bb_is_in_offset (bb, to) ||
+			((!r_list_contains(state, (void *)bb->jump) && anal_path_exists (core, bb->jump, to, bbs, depth - 1, state))) ||
+			((!r_list_contains(state, (void *)(bb->addr + bb->size)) && anal_path_exists (core, bb->addr + bb->size, to, bbs, depth - 1, state)))) {
+		r_list_prepend (bbs, bb);
+		return true;
+	}
+
+	// find our current function
 	r_list_foreach (core->anal->fcns, iter, fcn) {
-		if (!r_anal_fcn_is_in_offset (fcn, core->offset)) {
-			continue;
+		if (r_anal_fcn_in (fcn, from)) {
+			cur_fcn = fcn;
+			break;
 		}
-		r_list_foreach (fcn->bbs, iter2, bb) {
-			if (r_anal_bb_is_in_offset (bb, addr)) {
-				dest = bb;
-			}
-			if (r_anal_bb_is_in_offset (bb, core->offset)) {
-				root = bb;
-				r_list_append (list, list2);
+	}
+
+	// get call refs from current basic block and find a path from them
+	if (cur_fcn) {
+		refs = r_anal_fcn_get_refs (core->anal, cur_fcn);
+		if (refs) {
+			r_list_foreach (refs, iter, refi) {
+				if (refi->type == R_ANAL_REF_TYPE_CALL) {
+					if (r_anal_bb_is_in_offset (bb, refi->at)) {
+						if (refi->at != refi->addr && !r_list_contains(state, (void *)refi->addr) && anal_path_exists (core, refi->addr, to, bbs, depth - 1, state)) {
```

Comment:
void* an ut64 is a bad idea, also missing a space

Reply:
using the hashtable now.